README.md
# iBoardBot Variant with an Arduino UNO and any ESP8266 with AT Firmware

## Overview
This projects is in general a fork of the original repository of (jjrobots)[https://github.com/jjrobots/iBoardbot] which is not suitable for me
due to the lack of the Devian Control Board and/or a suitable atmega32u4 as replacement... but i got a bunch of old Arduino Nano and also some CNC Shields.
So i decided to use this as my control board for the iBoardBot.

Work is still in progress since i was in need to restructure the whole firmware and its still ongoing ;)


ChangeLog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0.2] - 2022-05-27
### Changed
- wifi ESP8266 arduino library

lib/wifiESP/wifiESP.h
#include <avr/pgmspace.h>
#include <Arduino.h>
// Needed libraries
#include <SoftwareSerial.h>

#define CNC_SHIELD_RX_PIN A5
#define CNC_SHIELD_TX_PIN A4
#define WIFI_SERIAL_SPEED 57600
#define WIFICONFIG_MAXLEN 30

#define SERVER_HOST "ibb.jjrobots.com"
#define SERVER_URL "http://ibb.jjrobots.com/ibbsvr/ibb.php"

class wifiESP
{
    private:
        //EEPROM object to store Wifi configuration
        struct WifiConfigS {
            uint8_t status;
            char ssid[WIFICONFIG_MAXLEN];
            char pass[WIFICONFIG_MAXLEN];
            char proxy[WIFICONFIG_MAXLEN];
            unsigned int port;
        };

        WifiConfigS WifiConfig;

        char MAC[13];  // MAC address of Wifi module

        // Global variables for message reading
        long message_timer_init;
        uint8_t message_readed;
        int message_index;
        int message_index_buffer;
        int message_size;
        bool message_chunked;
        uint8_t message_status;
        char mc0,mc1,mc2,mc3,mc4,mc5,mc6,mc7;  // Last 5 chars from message for parsing...
        unsigned char buffer[780];   // buffer to store incomming message from server

        void _flushSerial(void);
        uint8_t _readChar(char c[5]);
        void _wait(unsigned int timeout_secs);
        uint8_t _waitFor(const char *stopstr, unsigned int timeout_secs);
        uint8_t _waitFor2(const char *stopstr, const char *stopstr2, unsigned int timeout_secs);
        uint8_t _waitforMessage(uint8_t timeout_secs);
        uint8_t _extractGETParam(char *param, char ch);

    public:
        // public methods
        wifiESP(bool readConfig);
        ~wifiESP() {};
        
        void readConfigEEPROM(void);
        void writeWifiConfigEEPROM(uint8_t status, char ssid[30], char pass[30], char proxy[30], unsigned int port);
        void webServerConfig(void);
        void configWeb(uint8_t tcp_ch);
        void configWeb_advanced(uint8_t tcp_ch);
        void configWebOK(uint8_t tcp_ch);
        uint8_t getMac(char *MAC);
        uint8_t sendHTTP(char *url);
        bool connect(void);
        void wifiConfigurationMode(void);

        char* getSSID(void);
        char* getPASS(void);
        char* getProxy(void);
        unsigned int getPort(void);

        uint8_t getStatus(void);

        void setCredentials(char* ssid, char* pass);
        void setStatus(uint8_t status);
        void setPort(unsigned int port);
        void setProxy(char* proxy);

        uint8_t reset(void);
        void sendRaw(char msg);
        int available(void);
        int read(void);
};


include/configuration.h
#ifndef _CONFIGURATION_H_
#define _CONFIGURATION_H_
// iBoardBot Arduino UNO project configuration file
#include <Arduino.h>

// Pin configuration
#define STEPPER_EN        8
#define STEPPERX_STEP     2
#define STEPPERX_DIR      5
#define STEPPERY_STEP     4
#define STEPPERY_DIR      7
#define SERVO1_PIN        9
#define SERVO2_PIN        10
#define WIFI_CONFIG_PIN   A0

#define DEBUG 1

#define VERSION "iBoardBot Arduino UNO with CNC Shield v1.0.0.1"

// This depends on the pulley teeth and microstepping. For 20 teeth GT2 and 1/16 => 80  200*16 = 3200/(20*2mm) = 80
#define X_AXIS_STEPS_PER_UNIT 80    
// This depends on the pulley teeth and microstepping. For 42 teeth GT2 and 1/16 => 38  200*16 = 3200/(42*2mm) = 38
#define Y_AXIS_STEPS_PER_UNIT 38 

// Servo values por NoPaint, Paint and Erase servo positions...
#define SERVO1_PAINT 180
#define SERVO1_ERASER 105
#define SERVO2_LIFT 20
#define SERVO2_PAINT 120

#endif

src/main.cpp
#include <Arduino.h>
#include <stdint.h>

#include <BasicStepperDriver.h>
#include <A4988.h>
#include <Servo.h>
#include <wifiESP.h>

// Configuration
#include <configuration.h>

#define RPM 20
// Since microstepping is set externally, make sure this matches the selected mode
// If it doesn't, the motor will move at a different RPM than chosen
// 1=full step, 2=half step etc.
#define MICROSTEPS 16

//A4988 StepperX(X_AXIS_STEPS_PER_UNIT, STEPPERX_DIR, STEPPERX_STEP, STEPPER_EN);
//A4988 StepperY(Y_AXIS_STEPS_PER_UNIT, STEPPERY_DIR, STEPPERY_STEP, STEPPER_EN);
//Servo Servo1;
//Servo Servo2;

wifiESP Wifi(false);

// Wifi status
bool Wconnected;
uint8_t Network_errors;
char MAC[13];  // MAC address of Wifi module

void setup()
{
    pinMode(WIFI_CONFIG_PIN, INPUT); // WIFI CONFIG BUTTON => sensor imput
    digitalWrite(WIFI_CONFIG_PIN, HIGH); // Enable Pullup on A0

    //StepperX.begin(RPM, MICROSTEPS);
    //StepperY.begin(RPM, MICROSTEPS);
    // if using enable/disable on ENABLE pin (active LOW) instead of SLEEP uncomment next line
    //StepperX.setEnableActiveState(LOW);
    //StepperY.setEnableActiveState(LOW);

    //Servo1.attach(9);  // attaches the servos on pin 9 and 10 to the servo object
    //Servo2.attach(10);

    Serial.begin(115200); // Serial output
    
    //delay(1000);

    // Manual Configuration mode? => Erase stored Wifi parameters (force A0 to ground)
    //if (digitalRead(WIFI_CONFIG_PIN) == LOW)
    //  Wifi.writeWifiConfigEEPROM(0, "", "", "", 0);
    //Wifi.writeWifiConfigEEPROM(0,"","","",0); // Force config

    
  #ifdef DEBUG
    delay(1000);  // Only needed for serial debug
    Serial.println(VERSION);
  #endif

    Serial.println(F("Reading Wifi configuration..."));
    //Wifi.readConfigEEPROM();

    Serial.println(Wifi.getStatus());
    Serial.println(Wifi.getSSID());
    //Serial.println(WifiConfig.pass);
    Serial.println("****");
    Serial.println(Wifi.getProxy());
    Serial.println(Wifi.getPort());

    // if wifi parameters are not configured we start the config web server
    if (Wifi.getStatus() != 1) {
        Wifi.wifiConfigurationMode();
    }

    Wconnected = false;
    Network_errors = 0;
    while (!Wconnected) {

        Wifi.reset();
        //Wifi.getMac(MAC); // Todo. its just read from the incoming stream but didnt initiate it

        if (Wifi.connect()) {
            //digitalWrite(13, HIGH); // enable servos
            Wconnected = true;
        }
        else {
            Serial.println(F("Error connecting to network!"));
            digitalWrite(13, LOW);
            delay(2500);
            Wconnected = false;
            Network_errors++;
            // If connection fail 2 times we enter WifiConfigurationMode
            if (Network_errors >= 2) {
              Wifi.wifiConfigurationMode();
              Network_errors = 0;
            }
        }
    }
}

void loop()
{
}